### PR TITLE
Add opentelemetry trace instrumentation for DBT executions

### LINKED ISSUE #11367: [Feature] DBT Instrumentation with opentelemetry tracing
### Is this your first time submitting a feature request?

- [x] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [x] I have searched the existing issues, and I could not find an existing issue for this feature
- [x] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

We are collaborating to add opentelemetry auto instrumentation to DBT execution which involves in code changes to dbt-core, dbt-common, dbt-adapters. 

As part of this feature we want to achieve the following:
1. Add span to trace the dbt model execution.
2. Add span for pre_hooks and post_hooks executed within a model and the parent of these spans gonna be model span.
3. Add a span for on-run-start and on-run-end executions.
4. Add a span called metadata setup which traces the metadata queries DBT submits to find database/schema and create schema if required.

To support this we add new dependency on opentelemetry-api.

### Describe alternatives you've considered

_No response_

### Who will this benefit?

Snowflake customers can have auto-instrumented Telemetry for their dbt workloads but we believe this would be useful for other users of DBT to understand and debug the pipeline by configuring opentelemetry and collecting the exported instrumentation.

### Are you interested in contributing this feature?

Yes

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by sfc-gh-ppruett**
Resolves #11367

### Problem

Adds OpenTelemetry tracing to dbt-core's execution path and enriches the
emitted span data so traces align with the dbt-core 2.0 telemetry model. 

This change has a corresponding change in dbt-common ([PR](https://github.com/dbt-labs/dbt-common/pull/369))

### Solution

- Instrument node execution, on-run-start/end hooks, per-model pre/post hooks, and metadata cache setup with OpenTelemetry spans; propagate trace context into node-runner worker threads.
- Instrumentation is gated behind --snowflake-projects-otel CLI flag (default off; env var
DBT_ENGINE_SNOWFLAKE_PROJECTS_OTEL; also settable via dbt_project.yml flags).
- Represent DAG dependencies as span links from each node to its upstreams.
- Node spans: rename attribute keys to bare snake_case in alignment with dbt-core 2.0 (`node_outcome`, `materialization`, `database`, `schema`); add `unique_id`, `name`, `node_type`, `identifier`, `relative_path`, `rows_affected`, and `query_id`; set span status OK on success; skip attributes whose value is None so non-relational nodes (unit tests, exposures) don't emit empty attributes.
- Hook spans: emit a child span per hook (`hook_type`, `package_name`, `name`, `hook_index`, `unique_id`, `hook_outcome`, OK/ERROR status) under a per-type span.
- Instrument with the OpenTelemetry API only; the embedding process owns the `TracerProvider`/exporter (no exporter configured in-repo).

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me.
- [x] I have run this code in development, and it appears to resolve the stated issue.
- [x] This PR includes tests, or tests are not required or relevant for this PR.
- [x] This PR has no interface changes (e.g., macros, CLI, logs, JSON artifacts, config files, adapter interface, etc.) or this PR has already received feedback and approval from Product or DX.
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions.

### CODE REVIEW

**Comment by aahel** [comment]
> Thanks for your pull request, and welcome to our community! We require contributors to sign our [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScfOV7K4enYRHozrDRP6BBIXjOij-JDGca6WBTHyP_ANXSqlg/viewform) and we don't seem to have your signature on file. Check out [this article](https://docs.getdbt.com/docs/contributor-license-agreements) for more information on why we have a CLA.
> 
> In order for us to review and merge your code, please submit the Individual Contributor License Agreement form attached above above. If you have questions about the CLA, or if you believe you've received this message in error, please reach out through a comment on this PR.
> 
> CLA has not been signed by users: @sfc-gh-ppruett

@sfc-gh-ppruett

**Comment by aahel** [comment]
target branch should be 1.latest not 1.12.latest. changing it

File: core/dbt/task/runnable.py
Comment (tauhid621) [inline core/dbt/task/runnable.py:390]: We should trace the exception scenario here.

File: core/dbt/clients/jinja.py [hunk 1]
```diff
@@ -58,6 +60,9 @@ def __init__(
         super().__init__(macro, context)
         self.node = node
         self.stack = stack
+        # Tracer handle is created unconditionally; it is inert (creates no spans)
+        # until the --snowflake-projects-otel gate opens a span in __call__.
+        self.macro_tracer = trace.get_tracer("dbt.runner")
```
Comment (tauhid621) [inline core/dbt/clients/jinja.py]: I'm wondering if we should create this as a module level tracer to avoid creating it for all calls 🤔 
Something like `_TRACER = trace.get_tracer("dbt.runner")`

**Review by tauhid621 — COMMENTED** [review]
Thanks for the changes. They look good. 
Do we need a root span under which all spans would be created? 🤔

File: core/dbt/task/runnable.py
Comment (sfc-gh-ppruett) [reply core/dbt/task/runnable.py:390]: Good idea. Added.

File: core/dbt/clients/jinja.py (hunk 1 above)
Comment (sfc-gh-ppruett) [reply core/dbt/clients/jinja.py]: Made the change!

**Comment by sfc-gh-ppruett** [comment]
> Thanks for the changes. They look good. Do we need a root span under which all spans would be created? 🤔

Thanks, @tauhid621 . Good idea, I went ahead and added it. Also discovered a edge case related to the span nesting for build when unit tests were involved, so I fixed that as well.

**Review by tauhid621 — APPROVED** [review]
Thanks for addressing the feedback 🙇 
Could you please sign the cla as well.
